feat: added watsonX AI project automation, updated PDFs for ingestion

brendankellyibm opened the pull request:
### Description

- updated PDFs for data ingestion by Watson Discovery
- added WatsonX AI project configuration as we do not want to use the default created by the WatsonX SaaS DA
- created COS instance for that project
- created Continuous Delivery service so that toolchains can be run

### Release required?
<!--- Identify the type of release. For information about the changes in a semantic versioning release, see [Release versioning](https://terraform-ibm-modules.github.io/documentation/#/versioning). --->

- [ ] No release
- [ ] Patch release (`x.x.X`)
- [x] Minor release (`x.X.x`)
- [ ] Major release (`X.x.x`)

##### Release notes content

<!--- If a release is required, replace this text with information that users need to know about the release. Write the release notes to help users understand the changes, and include information about how to update from the previous version.

Your notes help the merger write the commit message for the PR that is published in the release notes for the module. --->

### Run the pipeline

If the CI pipeline doesn't run when you create the PR, the PR requires a user with GitHub collaborators access to run the pipeline.

Run the CI pipeline when the PR is ready for review and you expect tests to pass. Add a comment to the PR with the following text:

```
/run pipeline
```

### Checklist for reviewers

- [ ] If relevant, a test for the change is included or updated with this PR.
- [ ] If relevant, documentation for the change is included or updated with this PR.

### For mergers

- Use a conventional commit message to set the release level. Follow the [guidelines](https://terraform-ibm-modules.github.io/documentation/#/merging.md).
- Include information that users need to know about the PR in the commit message. The commit message becomes part of the GitHub release notes.
- Use the **Squash and merge** option.

ocofaigh commented on solutions/banking/main.tf:
```diff
@@ -15,51 +15,51 @@ resource "random_string" "unique_identifier" {
-  machine_learning_guid = ibm_resource_instance.machine_learning_instance.guid
-  machine_learning_crn  = ibm_resource_instance.machine_learning_instance.crn
-  machine_learning_name = ibm_resource_instance.machine_learning_instance.resource_name
+  source                = "git::git@github.com:terraform-ibm-modules/terraform-ibm-watsonx-saas-da.git//configure_project?ref=v0.2.0"
```
this should use a local path instead pulling from github, otherwise it will always be pulling an older version

ocofaigh replied:
ignore this - this is a different repo my bad

ocofaigh commented on solutions/banking/main.tf:
```diff
@@ -15,51 +15,51 @@ resource "random_string" "unique_identifier" {
+  # If an existing resource group is not set (null), then create a new one
+  resource_group_name          = var.resource_group_name == null ? local.unique_identifier : null
+  existing_resource_group_name = var.resource_group_name
+}
```
The approach other DAs have gone with is this:
```
module "resource_group" {
  source                       = "terraform-ibm-modules/resource-group/ibm"
  version                      = "1.1.5"
  resource_group_name          = var.use_existing_resource_group == false ? var.resource_group_name : null
  existing_resource_group_name = var.use_existing_resource_group == true ? var.resource_group_name : null
}
```

ocofaigh commented on solutions/banking/main.tf:
```diff
@@ -15,51 +15,51 @@ resource "random_string" "unique_identifier" {
+
+# create CD service for toolchain use
+resource "ibm_resource_instance" "cd_instance" {
+  name              = "${var.prefix}-cd-instance"
```
what happens if an account already has this service? Will this fail? Should we have a variable to opt in / out?

brendankellyibm replied:
I believe the CD service works on a per-resource group basis, not account. It's still a valid point - how would I check for an existing instance?

ocofaigh replied:
You can't, you would need the user to specify with a boolean if they want to create the service or not

ocofaigh commented on solutions/banking/variables.tf:
```diff
@@ -4,11 +4,27 @@ variable "ibmcloud_api_key" {
   sensitive   = true
 }
 
+variable "prefix" {
```
just a warning that we found consumes want full control over resource naming - if you don't really need a prefix, I would not add one here

ocofaigh reviewed (dismissed):
left some comments

rajatagarwal-ibm commented on solutions/banking/main.tf:
```diff
@@ -15,51 +15,51 @@ resource "random_string" "unique_identifier" {
+  source            = "terraform-ibm-modules/cos/ibm//modules/fscloud"
+  version           = "7.5.3"
+  resource_group_id = module.resource_group.resource_group_id
+  cos_instance_name = "gen-ai-rag-sample-app-cos-instance"
```
Could you add a `prefix` to the name to avoid conflicts?

ocofaigh replied:
I was suggesting to remove prefix variable and expose `cos_instance_name` as a variable (https://github.com/terraform-ibm-modules/terraform-ibm-rag-sample-da/pull/15#discussion_r1560786933)

brendankellyibm replied:
I'll add the prefix for now, to remain consistent with the SaaS DA.

rajatagarwal-ibm commented on solutions/banking/main.tf:
```diff
@@ -15,51 +15,51 @@ resource "random_string" "unique_identifier" {
+}
+
+# create CD service for toolchain use
+resource "ibm_resource_instance" "cd_instance" {
```
@brendankellyibm  the `cd_instance` is also created as part of the prereq for the tests where toolchains are also created. 

Do we need `cd_instance` here as part of the DA? We are not creating any toolchains in the DA itself.

ocofaigh replied:
Did we also reach out to ALM DA team to see why they don't support creating the service?

brendankellyibm replied:
I'm creating it here until we have an option to provision the CD service in the ALM DA - but in any case, they won't be doing it before Think, so we can revisit it afterwards.

@rajatagarwal-ibm it is required to have the CD service to run the toolchains created previously.

rajatagarwal-ibm commented on solutions/banking/main.tf:
```diff
@@ -15,51 +15,51 @@ resource "random_string" "unique_identifier" {
-  machine_learning_crn  = ibm_resource_instance.machine_learning_instance.crn
-  machine_learning_name = ibm_resource_instance.machine_learning_instance.resource_name
+  source                = "git::git@github.com:terraform-ibm-modules/terraform-ibm-watsonx-saas-da.git//configure_project?ref=v0.2.0"
+  project_name          = "RAG-sample-project"
```
Could you add a `prefix` to the name to avoid conflicts?

brendankellyibm replied:
Done in latest commit

rajatagarwal-ibm commented on tests/scripts/pre-validation.sh:
```diff
@@ -21,28 +22,45 @@ TF_VARS_FILE="terraform.tfvars"
+  watson_machine_learning_instance_guid_var_name="watson_machine_learning_instance_guid"
+  watson_machine_learning_instance_resource_name_var_name="watson_machine_learning_instance_resource_name"
 
+  resource_group_name_value=$(terraform output -state=terraform.tfstate -raw resource_group_name)
```
you need to add all of these new outputs to the "tests/resources/existing-resources/outputs.tf"

brendankellyibm replied:
I think this is done now

brendankellyibm commented:
@ocofaigh @rajatagarwal-ibm made some changes based on feedback 

- Made the CD service optional
- fixed outputs for tests
- updated resource group creation logic

ocofaigh reviewed (commented):
@rajatagarwal-ibm do you want to test the script updates by running catalog pipeline against this branch?

rajatagarwal-ibm commented:
@brendankellyibm @ocofaigh let me do catalog stuff against this branch, I will share findings.

rajatagarwal-ibm commented:
@brendankellyibm the validation failed with the following error

```
2024/04/12 09:47:33 Terraform init | Downloading registry.terraform.io/terraform-ibm-modules/cbr/ibm 1.19.1 for cos.instance_cbr_rules...
 2024/04/12 09:47:33 Terraform init | - cos.instance_cbr_rules in .terraform/modules/cos.instance_cbr_rules/modules/cbr-rule-module
 2024/04/12 09:47:33 Terraform init | Downloading registry.terraform.io/terraform-ibm-modules/resource-group/ibm 1.1.5 for resource_group...
 2024/04/12 09:47:41 Terraform init | - resource_group in .terraform/modules/resource_group
 2024/04/12 09:47:41 Terraform init | 
 2024/04/12 09:47:41 Terraform init | Error: Failed to download module
 2024/04/12 09:47:41 Terraform init | 
 2024/04/12 09:47:41 Terraform init |   on main.tf line 43:
 2024/04/12 09:47:41 Terraform init |   43: module "configure_project" {
 2024/04/12 09:47:41 Terraform init | 
 2024/04/12 09:47:41 Terraform init | Could not download module "configure_project" (main.tf:43) source code from
 2024/04/12 09:47:41 Terraform init | "git::ssh://git@github.com/terraform-ibm-modules/terraform-ibm-watsonx-saas-da.git?ref=v0.2.0":
 2024/04/12 09:47:41 Terraform init | error downloading
 2024/04/12 09:47:41 Terraform init | 'ssh://git@github.com/terraform-ibm-modules/terraform-ibm-watsonx-saas-da.git?ref=v0.2.0':
 2024/04/12 09:47:41 Terraform init | /usr/bin/git exited with 128: Cloning into
 2024/04/12 09:47:41 Terraform init | '.terraform/modules/configure_project'...
 2024/04/12 09:47:41 Terraform init | Host key verification failed.
 2024/04/12 09:47:41 Terraform init | fatal: Could not read from remote repository.
 2024/04/12 09:47:41 Terraform init | 
 2024/04/12 09:47:41 Terraform init | Please make sure you have the correct access rights
 2024/04/12 09:47:41 Terraform init | and the repository exists.
 2024/04/12 09:47:41 Terraform init | 
```

rajatagarwal-ibm commented:
I have found an issue with `ibm_catalog.json` file. Would you be able to fix it as part of this PR @brendankellyibm, please?

Replace the architecture block with this one

```
"architecture": {
              "diagrams": [
                {
                  "diagram": {
                    "url": "https://raw.githubusercontent.com/terraform-ibm-modules/terraform-ibm-rag-sample-da/main/reference-architecture/ce_da_architecture_diagram.drawio.svg",
                    "caption": "RAG DA reference architecture diagram",
                    "type": "image/svg+xml"
                  },
                  "description": "RAG DA reference architecture diagram"
                }
              ]
            },
```

brendankellyibm commented:
@rajatagarwal-ibm I've updated ibm_catalog.json, and changed the source of the module to https git.

vburckhardt commented:
/run pipeline

rajatagarwal-ibm commented:
@brendankellyibm CRA scan failed, could you add those required values to the pre-validation script?

rajatagarwal-ibm commented:
validation is failing with the following error

```
 2024/04/12 11:49:10 Terraform plan |  Error: Missing required argument
 2024/04/12 11:49:10 Terraform plan | 
 2024/04/12 11:49:10 Terraform plan | The argument "uri" is required, but was not set.

```

@ocofaigh do you have any idea on this?

ocofaigh commented:
@rajatagarwal-ibm this generally means there is no provider config

vburckhardt commented:
/run pipeline

rajatagarwal-ibm commented:
/run pipeline

rajatagarwal-ibm commented:
/run pipeline

rajatagarwal-ibm commented:
/run pipeline

rajatagarwal-ibm commented:
/run pipeline

rajatagarwal-ibm commented:
/run pipeline

rajatagarwal-ibm commented:
/run pipeline

rajatagarwal-ibm commented:
/run pipeline

rajatagarwal-ibm commented:
/run pipeline

terraform-ibm-modules-ops commented:
:tada: This PR is included in version 1.1.0 :tada:

The release is available on:
- [GitHub release](https://github.com/terraform-ibm-modules/terraform-ibm-rag-sample-da/releases/tag/v1.1.0)
- `v1.1.0`

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
